File: calorie_calc.js
#!/usr/bin/env node
/**
 * calorie_calc.js - Калькулятор калорий на JavaScript (Node.js CLI)
 */
const fs = require('fs');
const { program } = require('commander');

const ACTIVITY_LEVELS = {
    sedentary: 1.2,
    light: 1.375,
    moderate: 1.55,
    active: 1.725,
    very_active: 1.9,
};
const ACTIVITY_NAMES = {
    sedentary: 'Сидячий образ жизни',
    light: 'Лёгкая активность (1-3 дня/нед)',
    moderate: 'Умеренная активность (3-5 дней/нед)',
    active: 'Высокая активность (6-7 дней/нед)',
    very_active: 'Очень высокая активность (тяжёлая работа/спорт)',
};
const GOALS = ['maintain', 'lose', 'gain'];
const GOAL_NAMES = {
    maintain: 'Поддержание веса',
    lose: 'Похудение (дефицит 15%)',
    gain: 'Набор массы (профицит 15%)',
};
const GENDER_NAMES = { male: 'Мужской', female: 'Женский' };

function bmrMifflin(gender, age, height, weight) {
    if (gender === 'male') {
        return 10 * weight + 6.25 * height - 5 * age + 5;
    } else {
        return 10 * weight + 6.25 * height - 5 * age - 161;
    }
}

function bmrHarris(gender, age, height, weight) {
    if (gender === 'male') {
        return 88.362 + 13.397 * weight + 4.799 * height - 5.677 * age;
    } else {
        return 447.593 + 9.247 * weight + 3.098 * height - 4.330 * age;
    }
}

function calculateTdee(bmr, activity) {
    return bmr * ACTIVITY_LEVELS[activity];
}

function calculateGoalCalories(tdee, goal) {
    if (goal === 'maintain') return tdee;
    if (goal === 'lose') return tdee * 0.85;
    if (goal === 'gain') return tdee * 1.15;
    return tdee;
}

function calculateMacros(calories, goal) {
    let proteinRatio = 0.30, fatRatio = 0.25, carbRatio = 0.45;
    if (goal === 'lose') {
        proteinRatio = 0.35; fatRatio = 0.25; carbRatio = 0.40;
    } else if (goal === 'gain') {
        proteinRatio = 0.30; fatRatio = 0.20; carbRatio = 0.50;
    }
    return {
        protein: (calories * proteinRatio) / 4,
        fat: (calories * fatRatio) / 9,
        carbs: (calories * carbRatio) / 4,
    };
}

function generateReport(params, bmr, tdee, goalCal, macros) {
    const lines = [];
    lines.push('='.repeat(50));
    lines.push('📋 ОТЧЁТ ПО КАЛОРИЯМ');
    lines.push('='.repeat(50));
    lines.push(`Пол: ${GENDER_NAMES[params.gender]}`);
    lines.push(`Возраст: ${params.age} лет`);
    lines.push(`Рост: ${params.height} см`);
    lines.push(`Вес: ${params.weight} кг`);
    lines.push(`Уровень активности: ${ACTIVITY_NAMES[params.activity]}`);
    lines.push(`Цель: ${GOAL_NAMES[params.goal]}`);
    lines.push('-'.repeat(50));
    lines.push(`BMR (базовый метаболизм): ${bmr.toFixed(1)} ккал/день`);
    lines.push(`TDEE (суточная норма): ${tdee.toFixed(1)} ккал/день`);
    lines.push(`Рекомендуемая калорийность для цели: ${goalCal.toFixed(1)} ккал/день`);
    lines.push('-'.repeat(50));
    lines.push('🍽️ РАСПРЕДЕЛЕНИЕ ПО МАКРОНУТРИЕНТАМ:');
    lines.push(`Белки: ${macros.protein.toFixed(1)} г (${((macros.protein*4/goalCal)*100).toFixed(0)}%)`);
    lines.push(`Жиры: ${macros.fat.toFixed(1)} г (${((macros.fat*9/goalCal)*100).toFixed(0)}%)`);
    lines.push(`Углеводы: ${macros.carbs.toFixed(1)} г (${((macros.carbs*4/goalCal)*100).toFixed(0)}%)`);
    lines.push('='.repeat(50));
    return lines.join('\n');
}

program
    .option('-g, --gender <gender>', 'Пол (male/female)')
    .option('-a, --age <age>', 'Возраст (лет)', parseFloat)
    .option('--height <height>', 'Рост (см)')
    .option('--height-ft <feet>', 'Рост (футы)', parseInt)
    .option('--height-in <inches>', 'Рост (дюймы)', parseInt, 0)
    .option('--weight <weight>', 'Вес (кг)')
    .option('--weight-lb <weight>', 'Вес (фунты)')
    .option('--activity <activity>', 'Уровень активности', /^(sedentary|light|moderate|active|very_active)$/i)
    .option('--goal <goal>', 'Цель (maintain/lose/gain)', 'maintain')
    .option('--formula <formula>', 'Формула (msj/hb)', 'msj')
    .option('--output <file>', 'Сохранить отчёт в файл')
    .parse(process.argv);

const opts = program.opts();

if (!opts.gender || !opts.age || !opts.activity) {
    console.error('Укажите --gender, --age, --activity');
    process.exit(1);
}

let height = opts.height;
let weight = opts.weight;
if (opts.heightFt !== undefined) {
    height = opts.heightFt * 30.48 + (opts.heightIn || 0) * 2.54;
}
if (opts.weightLb !== undefined) {
    weight = opts.weightLb * 0.453592;
}
if (height === undefined || weight === undefined) {
    console.error('Укажите рост и вес (метрические или имперские)');
    process.exit(1);
}

const params = {
    gender: opts.gender.toLowerCase(),
    age: opts.age,
    height: height,
    weight: weight,
    activity: opts.activity.toLowerCase(),
    goal: opts.goal.toLowerCase(),
    formula: opts.formula.toLowerCase(),
};

let bmr;
if (params.formula === 'msj') {
    bmr = bmrMifflin(params.gender, params.age, params.height, params.weight);
} else {
    bmr = bmrHarris(params.gender, params.age, params.height, params.weight);
}
const tdee = calculateTdee(bmr, params.activity);
const goalCal = calculateGoalCalories(tdee, params.goal);
const macros = calculateMacros(goalCal, params.goal);

const report = generateReport(params, bmr, tdee, goalCal, macros);
console.log(report);

if (opts.output) {
    fs.writeFileSync(opts.output, report, 'utf8');
    console.log(`Отчёт сохранён в ${opts.output}`);
}

File: calorie_calc.go
// calorie_calc.go - Калькулятор калорий на Go (CLI)
package main

import (
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var activityLevels = map[string]float64{
	"sedentary":  1.2,
	"light":      1.375,
	"moderate":   1.55,
	"active":     1.725,
	"very_active": 1.9,
}
var activityNames = map[string]string{
	"sedentary":   "Сидячий образ жизни",
	"light":       "Лёгкая активность (1-3 дня/нед)",
	"moderate":    "Умеренная активность (3-5 дней/нед)",
	"active":      "Высокая активность (6-7 дней/нед)",
	"very_active": "Очень высокая активность (тяжёлая работа/спорт)",
}
var goalNames = map[string]string{
	"maintain": "Поддержание веса",
	"lose":     "Похудение (дефицит 15%)",
	"gain":     "Набор массы (профицит 15%)",
}
var genderNames = map[string]string{
	"male":   "Мужской",
	"female": "Женский",
}

func bmrMifflin(gender string, age, height, weight float64) float64 {
	if gender == "male" {
		return 10*weight + 6.25*height - 5*age + 5
	}
	return 10*weight + 6.25*height - 5*age - 161
}

func bmrHarris(gender string, age, height, weight float64) float64 {
	if gender == "male" {
		return 88.362 + 13.397*weight + 4.799*height - 5.677*age
	}
	return 447.593 + 9.247*weight + 3.098*height - 4.330*age
}

func calculateTdee(bmr float64, activity string) float64 {
	return bmr * activityLevels[activity]
}

func calculateGoalCalories(tdee float64, goal string) float64 {
	switch goal {
	case "maintain":
		return tdee
	case "lose":
		return tdee * 0.85
	case "gain":
		return tdee * 1.15
	default:
		return tdee
	}
}

func calculateMacros(calories float64, goal string) (protein, fat, carbs float64) {
	proteinRatio, fatRatio, carbRatio := 0.30, 0.25, 0.45
	switch goal {
	case "lose":
		proteinRatio, fatRatio, carbRatio = 0.35, 0.25, 0.40
	case "gain":
		proteinRatio, fatRatio, carbRatio = 0.30, 0.20, 0.50
	}
	protein = (calories * proteinRatio) / 4
	fat = (calories * fatRatio) / 9
	carbs = (calories * carbRatio) / 4
	return
}

func generateReport(gender, age, height, weight, activity, goal, formula string, bmr, tdee, goalCal, protein, fat, carbs float64) string {
	lines := []string{}
	lines = append(lines, strings.Repeat("=", 50))
	lines = append(lines, "📋 ОТЧЁТ ПО КАЛОРИЯМ")
	lines = append(lines, strings.Repeat("=", 50))
	lines = append(lines, fmt.Sprintf("Пол: %s", genderNames[gender]))
	lines = append(lines, fmt.Sprintf("Возраст: %.0f лет", age))
	lines = append(lines, fmt.Sprintf("Рост: %.1f см", height))
	lines = append(lines, fmt.Sprintf("Вес: %.1f кг", weight))
	lines = append(lines, fmt.Sprintf("Уровень активности: %s", activityNames[activity]))
	lines = append(lines, fmt.Sprintf("Цель: %s", goalNames[goal]))
	lines = append(lines, strings.Repeat("-", 50))
	lines = append(lines, fmt.Sprintf("BMR (базовый метаболизм): %.1f ккал/день", bmr))
	lines = append(lines, fmt.Sprintf("TDEE (суточная норма): %.1f ккал/день", tdee))
	lines = append(lines, fmt.Sprintf("Рекомендуемая калорийность для цели: %.1f ккал/день", goalCal))
	lines = append(lines, strings.Repeat("-", 50))
	lines = append(lines, "🍽️ РАСПРЕДЕЛЕНИЕ ПО МАКРОНУТРИЕНТАМ:")
	lines = append(lines, fmt.Sprintf("Белки: %.1f г (%.0f%%)", protein, (protein*4/goalCal)*100))
	lines = append(lines, fmt.Sprintf("Жиры: %.1f г (%.0f%%)", fat, (fat*9/goalCal)*100))
	lines = append(lines, fmt.Sprintf("Углеводы: %.1f г (%.0f%%)", carbs, (carbs*4/goalCal)*100))
	lines = append(lines, strings.Repeat("=", 50))
	return strings.Join(lines, "\n")
}

func main() {
	var (
		gender    string
		age       float64
		height    float64
		heightFt  int
		heightIn  int
		weight    float64
		weightLb  float64
		activity  string
		goal      string
		formula   string
		output    string
		useMetric bool
	)
	flag.StringVar(&gender, "gender", "", "Пол (male/female)")
	flag.Float64Var(&age, "age", 0, "Возраст (лет)")
	flag.Float64Var(&height, "height", 0, "Рост (см)")
	flag.IntVar(&heightFt, "height-ft", 0, "Рост (футы)")
	flag.IntVar(&heightIn, "height-in", 0, "Рост (дюймы)")
	flag.Float64Var(&weight, "weight", 0, "Вес (кг)")
	flag.Float64Var(&weightLb, "weight-lb", 0, "Вес (фунты)")
	flag.StringVar(&activity, "activity", "", "Уровень активности (sedentary,light,moderate,active,very_active)")
	flag.StringVar(&goal, "goal", "maintain", "Цель (maintain,lose,gain)")
	flag.StringVar(&formula, "formula", "msj", "Формула (msj/hb)")
	flag.StringVar(&output, "output", "", "Сохранить отчёт в файл")
	flag.Parse()

	if gender == "" || age == 0 || activity == "" {
		fmt.Println("Укажите --gender, --age, --activity")
		return
	}
	if height == 0 && (heightFt == 0 && heightIn == 0) {
		fmt.Println("Укажите рост (--height или --height-ft/--height-in)")
		return
	}
	if weight == 0 && weightLb == 0 {
		fmt.Println("Укажите вес (--weight или --weight-lb)")
		return
	}

	if height == 0 {
		height = float64(heightFt)*30.48 + float64(heightIn)*2.54
	}
	if weight == 0 {
		weight = weightLb * 0.453592
	}

	var bmr float64
	if formula == "msj" {
		bmr = bmrMifflin(gender, age, height, weight)
	} else {
		bmr = bmrHarris(gender, age, height, weight)
	}
	tdee := calculateTdee(bmr, activity)
	goalCal := calculateGoalCalories(tdee, goal)
	protein, fat, carbs := calculateMacros(goalCal, goal)

	report := generateReport(gender, age, height, weight, activity, goal, formula, bmr, tdee, goalCal, protein, fat, carbs)
	fmt.Println(report)

	if output != "" {
		os.WriteFile(output, []byte(report), 0644)
		fmt.Printf("Отчёт сохранён в %s\n", output)
	}
}
